infrastructure/src/test/java/org/corfudb/infrastructure/CorfuSequencerServerTest.java

```diff
@@ -0,0 +1,528 @@
+
+        // Assert that the payload has a BootstrapSequencerResponseMsg.NACK message
+        // and that the base header fields have remained the same
+        assertThat(compareBaseHeaderFields(request.getHeader(), response.getHeader())).isTrue();
+        assertThat(response.getPayload().hasBootstrapSequencerResponse()).isTrue();
+        assertThat(response.getPayload().getBootstrapSequencerResponse().getRespType())
+                .isEqualTo(Sequencer.BootstrapSequencerResponseMsg.Type.NACK);
+    }
+
+    /**
+     * Test that stale bootstrap requests are discarded.
+     * If epoch sent by the client is not greater than the sequencerEpoch then the sequencer should discard the
+     * BootstrapSequencerRequestMsg, and return a Sequencer.BootstrapSequencerResponseMsg.NACK message.
+     */
+    @Test
+    public void testStaleBootstrapRequest() {
+        sequencerServer = new SequencerServer(mockServerContext, spyFactoryHelper);
+        sequencerServer.setSequencerEpoch(0L);
+        RequestMsg request = getRequestMsg(
+                getBasicHeader(false, true),
+                CorfuProtocolSequencer.getBootstrapSequencerRequestMsg(
+                        Collections.emptyMap(),
+                        0,
+                        0,
+                        false)
+        );
+
+        when(mockServerContext.getServerEpoch()).thenReturn(1L);
+
+        // Invoke the handler method
+        sequencerServer.handleMessage(request, mockChannelHandlerContext, mockServerRouter);
+
+        // Verify that sendResponse() was called and capture the response object
+        ArgumentCaptor<ResponseMsg> responseCaptor = ArgumentCaptor.forClass(ResponseMsg.class);
+        verify(mockServerRouter)
+                .sendResponse(responseCaptor.capture(), any(ChannelHandlerContext.class));
+        ResponseMsg response = responseCaptor.getValue();
+
+        // Assert that the payload has a BootstrapSequencerResponseMsg.NACK message
+        // and that the base header fields have remained the same
+        assertThat(compareBaseHeaderFields(request.getHeader(), response.getHeader())).isTrue();
+        assertThat(response.getPayload().hasBootstrapSequencerResponse()).isTrue();
+        assertThat(response.getPayload().getBootstrapSequencerResponse().getRespType())
+                .isEqualTo(Sequencer.BootstrapSequencerResponseMsg.Type.NACK);
+    }
+
+    /**
+     * Test that the SequencerServer correctly handles a BootstrapSequencerRequestMsg. If the
+     * epoch sent by the client is greater than or equal to the current server epoch, and bootstrapWithoutTailsUpdate
+     * is set as false, a BootstrapSequencerResponseMsg.ACK message should be received.
+     * <p>
+     * Further, we request the server with a StreamsAddressRequestMsg and verify that
+     * the globalLogTail, trimMark and the streamsAddressMap were correctly updated on the server side.
+     */
+    @Test
+    public void testBootStrapSequencer() {
```

Comment:
Should we also have a test that validates that a bootstrap request with an older epoch is rejected?